text/0000-nested-publish.md

```diff
@@ -0,0 +1,186 @@
+- Feature Name: `nested_publish`
+- Start Date: 2023-06-30
+- RFC PR: [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452)
+- Rust Issue: ...
+
+# Summary
+[summary]: #summary
+
+Allow Cargo packages to be bundled within other Cargo packages when they are published (not just in unpublished workspaces).
+
+# Motivation
+[motivation]: #motivation
+
+There are a number of reasons why a Rust developer currently may feel the need to create multiple library crates, and therefore multiple Cargo packages (since one package contains at most one library crate). These multiple libraries could be:
+
+* A trait declaration and a corresponding derive macro (which must be defined in a separate proc-macro library).
+* A library that uses a build script that uses another library or binary (e.g. for precomputation or bindings generation).
+* A logically singular library broken into multiple parts to speed up compilation.
+
+Currently, developers must publish these packages separately. This has several disadvantages (see the [Rationale](#rationale-and-alternatives) section for further details):
+
+* Clutters the public view of the registry with packages not intended to be usable on their own, and which may even become obsolete as internal architecture changes.
+* Requires multiple `cargo publish` operations (this could be fixed with bulk publication) and writing public metadata for each package.
+* Can result in semver violations and thus compilation failures, due to the developer not thinking about semver compatibility within the group.
+
+This RFC will allow developers to avoid all of these inconveniences and hazards by publishing a single package.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default (and always, prior to this RFC's implementation):
+
+* If your package contains any sub-packages, Cargo [excludes](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) them from the `.crate` archive file produced by `cargo package` and `cargo publish`.
+* If your package contains any non-`dev` dependencies which do not give a `version = "..."`, it cannot be published to `crates.io`.
+* If your package contains `[dev-dependencies]` which do not give a `version = "..."`, they are stripped out on publication.
+
+(By “**sub-package**” we mean a package (directory with `Cargo.toml`) which is a subdirectory of another package. We shall call the outermost such package, the package being published, the “**parent package**”.)
+
+You can change this default by placing in the manifest (`Cargo.toml`) of a sub-package:
+
+```toml
+[package]
+publish = "nested"
```

Comment:
Bikeshedding, but I feel like there could/should be a better way to indicate this.

Reply:
Can you say more about what would make a solution better? I think it is an important safety feature that a nested package is declared ineligible for standalone publication at the same time it is made eligible for nesting, to avoid unintended duplication (though this _could_ be done by checking parent directories instead).

Reply:
Honestly, I'm not sure what I'd like it to look like. It just _feels_ less than ideal. I know that isn't very helpful, admittedly.

Reply:
It could be on the dependency like in the blog post I linked elsewhere.

Reply:
This is a good alternative to have considered, but I don't think it is wise. I have added the following text to Alternatives explaining:

> Similar to the previous, we could allow marking specific `[dependencies]` as “include this for publication”. However, this would make it impossible to publish a nested package which is not a dependency of the parent package, which blocks the future possibility of public targets in nested packages (particularly, installable binary targets in sub-packages, which will frequently depend on the parent and not vice versa).

(I haven't yet gotten to reading those posts.)